zcash_keys/src/keys.rs

```diff
@@ -498,22 +505,28 @@ pub struct UnifiedAddressRequest {
     has_orchard: bool,
     has_sapling: bool,
     has_p2pkh: bool,
+    expiry_height: Option<BlockHeight>,
+    expiry_time: Option<u64>,
 }
 
 impl UnifiedAddressRequest {
-    /// Construct a new unified address request from its constituent parts.
-    ///
-    /// Returns `None` if the resulting unified address would not include at least one shielded receiver.
-    pub fn new(has_orchard: bool, has_sapling: bool, has_p2pkh: bool) -> Option<Self> {
-        let has_shielded_receiver = has_orchard || has_sapling;
-
-        if !has_shielded_receiver {
+    /// Construct a new unified address request from its constituent parts
+    pub fn new(
+        has_orchard: bool,
+        has_sapling: bool,
+        has_p2pkh: bool,
+        expiry_height: Option<BlockHeight>,
+        expiry_time: Option<u64>,
+    ) -> Option<Self> {
+        if !(has_sapling || has_orchard || has_p2pkh) {
             None
         } else {
             Some(Self {
                 has_orchard,
                 has_sapling,
                 has_p2pkh,
+                expiry_height,
+                expiry_time,
             })
         }
```

Comment:
```suggestion
        if has_sapling || has_orchard || (has_p2pkh && (expiry_height.is_some() || expiry_time.is_some())) {
            Some(Self {
                has_orchard,
                has_sapling,
                has_p2pkh,
                expiry_height,
                expiry_time,
            })
        } else {
            None
        }
```

Reply:
I didn't think that we were _requiring_ must-understand metadata to accompany p2pkh addresses? I don't see that anywhere in the current text of ZIP 316; https://zips.z.cash/zip-0316#requirements-for-both-unified-addresses-and-unified-viewing-keys just says that the "at least one shielded receiver" requirement is dropped.

Reply:
I thought we decided to make the F4Jumble input size 40 bytes so that the original rationale for excluding bare P2PKH addresses as UAs — that they can be more compatibly expressed as t-addresses — would be maintained? 

https://github.com/zcash/zips/pull/797/files#diff-8ff1060cb8fa47eb429fb6c849f1364e28fd4b42868ae8f56dd20ea8b9028b2c

Reply:
Addressed in 673e1a21d28209103dce50ede358491e5978d007, but I hope that it's clear how much less maintainable and error-prone this approach is.